proposals/NNNN-Keyboard-Enhancements.md

```diff
@@ -0,0 +1,219 @@
+        <description>Enumeration listing possible input character masking.</description>
+        <element name="ENABLE_INPUT_KEY_MASK" />
+        <element name="DISABLE_INPUT_KEY_MASK" />
+        <element name="USER_CHOICE_INPUT_KEY_MASK" />
+    </enum>    
+```
+
+#### Changes in _"KeyboardEvent"_ Enum
+
+If App sends _"maskInputCharacters"_ value as _"USER_CHOICE_INPUT_KEY_MASK"_, the HMI keyboard will provide a key, so that the user can unmask input characters.
+The app will be notified whether the input is masked or not, using _"UI.OnKeyboardInput"_ with below keyboard events.
+
+```xml
+<enum name="KeyboardEvent">
+	:
+	<element name="INPUT_KEY_MASK_ENABLED"/>
+	<element name="INPUT_KEY_MASK_DISABLED"/>
+</enum>
+```
+
+#### Addition of _"ConfigurableKeyboards"_ Struct
+
+This struct object will be returned with _"KeyboardCapabilities"_ in _"RegisterAppInterface"_ response. Each layout can have a different number of customizable buttons. Hence, this struct object will map a number of customizable buttons and keyboard layout.
+
+```xml
+    <struct name="ConfigurableKeyboards" since="X.X">
+        <description>
+            Describes number of cofigurable Keys for Special characters.
+        </description>
+        <param name="keyboardLayout" type="KeyboardLayout" mandatory="true"/>
+        <param name="numOfKeys" type="Integer" mandatory="true"/>
+    </struct>
+```
+
+
+
+#### Change in _"KeyboardLayout"_ enum
+
+```xml
+<enum name="KeyboardLayout">
+	:
+	<element name="NUMERIC"/>
+</enum>
+```
+
+#### Change in _"KeyboardProperties"_ struct
+
+1. App can request masking input characters using _"maskInputCharacters"_ param in _"KeyboardProperties"_ struct.
+2. App can request changes in special characters shown in the keyboard layout using _"customizeKeys"_ array. To set these buttons, app can just send an array of _"Strings"_ with special characters in _"customizeKeys"_ in _"SetGlobalProperties"_ RPC.
+3. If App does not send this string, HMI should show default special characters in the keyboard layout.
+4. If the number of keys in _"customizeKeys"_ array is more than customizable keys allowed, the SDL core should respond with _"INVALID_DATA"_ and the info string should include a detailed message, that **_"customizeKeys exceeds the number of customizable keys in this Layout"_**.
+5. If the number of keys in _"customizeKeys"_ array is less than or equal to customizable keys allowed, the SDL core should respond with _"SUCCESS"_. 
+6. HMI should show default characters in the remaining customizable Keys. if _"customizeKeys"_ array is less than or equal to the customizable keys allowed.
+7. If a certain special character is not supported by the system. The HMI should send _"WARNING"_ response, with _"info"_ text as _" some symbols might not be supported by system"_. This will give a chance to use symbols that are supported and also inform the app about the system not supporting certain characters.

```

Comment:
```suggestion
7. If a certain special character is not supported by the system the HMI should send _"WARNING"_ response, with _"info"_ text as _" some symbols might not be supported by system"_. This will give a chance to use symbols that are supported and also inform the app about the system not supporting certain characters.
```